endpoints/put.go

```diff
@@ -32,113 +31,199 @@ func NewPutHandler(backend backends.Backend, maxNumValues int, allowKeys bool) f
+					<-done
+				}
+			}
+		}
+
+		if exitData.status != http.StatusOK {
+			http.Error(w, exitData.errMsg, exitData.status)
 			return
 		}
 
-		resps := putResponsePool.Get().(*PutResponse)
-		resps.Responses = make([]PutResponseObject, len(put.Puts))
-		defer putResponsePool.Put(resps)
+		bytes, err := json.Marshal(backend.resp)
+		if err != nil {
+			http.Error(w, "Failed to serialize UUIDs into JSON.", http.StatusInternalServerError)
+			return
+		}
 
-		for i, p := range put.Puts {
-			if len(p.Value) == 0 {
-				http.Error(w, "Missing value.", http.StatusBadRequest)
-				return
-			}
-			if p.TTLSeconds < 0 {
-				http.Error(w, fmt.Sprintf("request.puts[%d].ttlseconds must not be negative.", p.TTLSeconds), http.StatusBadRequest)
+		/* Handles POST */
+		w.Header().Set("Content-Type", "application/json")
+		w.Write(bytes)
+	}
+}
+
+func validateAndEncode(puts *PutRequest, resp *PutResponse, exit *exitInfo) {
+	for index, p := range puts.Puts {
+		var toCache string
+
+		if len(p.Value) == 0 {
+			exit.errMsg = "Missing value."
+			exit.status = http.StatusBadRequest
+			return
+		}
+		if p.TTLSeconds < 0 {
+			exit.errMsg = fmt.Sprintf("request.puts[%d].ttlseconds must not be negative.", p.TTLSeconds)
+			exit.status = http.StatusBadRequest
+			return
+		}
+
+		if p.Type == backends.XML_PREFIX {
+			if p.Value[0] != byte('"') || p.Value[len(p.Value)-1] != byte('"') {
+				exit.errMsg = fmt.Sprintf("XML messages must have a String value. Found %v", p.Value)
+				exit.status = http.StatusBadRequest
 				return
 			}
 
-			var toCache string
-			if p.Type == backends.XML_PREFIX {
-				if p.Value[0] != byte('"') || p.Value[len(p.Value)-1] != byte('"') {
-					http.Error(w, fmt.Sprintf("XML messages must have a String value. Found %v", p.Value), http.StatusBadRequest)
-					return
-				}
+			// Be careful about the the cross-script escaping issues here. JSON requires quotation marks to be escaped,
+			// for example... so we'll need to un-escape it before we consider it to be XML content.
+			var interpreted string
+			json.Unmarshal(p.Value, &interpreted)
+			toCache = p.Type + interpreted
+		} else if p.Type == backends.JSON_PREFIX {
+			toCache = p.Type + string(p.Value)
+		} else {
+			exit.errMsg = fmt.Sprintf("Type must be one of [\"json\", \"xml\"]. Found %v", p.Type)
+			exit.status = http.StatusBadRequest
+			return
+		}
+		logrus.Debugf("Storing value: %s", toCache)
+		u2, err := uuid.NewV4()
+		if err != nil {
+			exit.errMsg = "Error generating version 4 UUID"
+			exit.status = http.StatusInternalServerError
+			return
+		}
+		resp.Responses[index].UUID = u2.String()
+		resp.toCacheStrings[index] = toCache
+	}
+	return
+}
+
+type backendCallObject struct {
+	client    backends.Backend
+	put       *PutRequest
+	resp      *PutResponse
+	allowKeys bool
+	ctx       context.Context
+}
+
+func callBackendInParallel(backend *backendCallObject, exit *exitInfo, indexChannel <-chan int, done chan<- bool) {
+	for index := range indexChannel {
```

Comment:
I'm not sure this provides parallelization. Instead of passing the indexChannel here, consider calling this method in each iteration of the callsite.

Something like:
```
func callBackend(backend *backendCallObject, exit *exitInfo, index int, done chan<- bool) {
  defer func() { done <- true }
  callBackendGet(backend, index)
  callBackendPut(backend, exit, index)
}
```

.. and at the call site ...

```
// start workers
for i, _ := range backend.put.Puts {
   go callBackend(backend, exitData, i, done)
}

// wait for all workers to be done
for _ , _ := range backend.put.Puts {
  <-done
}
```